### PR TITLE
Exchange trade

### PR DESCRIPTION
**Pull request by gyurka00**
Trade history

### CODE REVIEW

File: src/modules/exchangeProvider.js [hunk 1]
```diff
@@ -60,16 +61,22 @@ const onExchangeContractChange = (newVal, oldVal, objectPath) => {
...
 };
 
 // event NewOrder(uint indexed orderId, address indexed maker, uint price, uint tokenAmount, uint weiAmount);
-const onNewOrder = (orderId, maker, price, tokenAmount, weiAmount) => {
+const onNewOrder = function(orderId, maker, price, tokenAmount, weiAmount) {
     console.debug("exchangeProvider.onNewOrder: dispatching refreshExchange() and refreshOrders()");
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
+    store.dispatch(processNewTrade('NewOrder', userAccount, this));
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: this only needs to be dispatched if userAccount === maker, otherwise we display the trades of other accounts?

File: src/modules/exchangeProvider.js [hunk 2]
```diff
@@ -80,11 +87,12 @@ const onNewOrder = (orderId, maker, price, tokenAmount, weiAmount) => {
 };
 
 // CancelledOrder(uint indexed orderId, address indexed maker, uint tokenAmount, uint weiAmount);
-const onCancelledOrder = (orderId, maker, tokenAmount, weiAmount) => {
+const onCancelledOrder = function(orderId, maker, tokenAmount, weiAmount) {
     console.debug("exchangeProvider.onNewOrder: dispatching refreshExchange() and refreshOrders()");
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
+    store.dispatch(processNewTrade('CancelledOrder', userAccount, this));
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: same here, only dispatch if userAccount === maker

File: src/modules/exchangeProvider.js [hunk 3]
```diff
@@ -94,13 +102,14 @@ const onCancelledOrder = (orderId, maker, tokenAmount, weiAmount) => {
...
 // OrderFill(address indexed tokenBuyer, address indexed tokenSeller, uint buyTokenOrderId, uint sellTokenOrderId, uint price, uint weiAmount, uint tokenAmount);
-const onOrderFill = (tokenBuyer, tokenSeller, buyTokenOrderId, sellTokenOrderId, price, weiAmount, tokenAmount) => {
+const onOrderFill = function(tokenBuyer, tokenSeller, buyTokenOrderId, sellTokenOrderId, price, weiAmount, tokenAmount) {
     console.debug("exchangeProvider.onOrderFill: dispatching refreshExchange() and regreshOrders()");
     // FIXME: shouldn't do refresh for each orderFill event becuase multiple orderFills emmitted for one new order
     //          but newOrder is not emmited when a sell fully covered by orders and
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
+    store.dispatch(processNewTrade('OrderFill', userAccount, this));
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: same here, only dispatch if userAccount === maker or userAccount === taker
if you dispatch twice when maker === taker, then it will show up twice in trades history (which is expected if the match is with myself)

File: src/containers/exchange/components/TradeBook.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+import React from "react";
+
+import { Pblock } from "components/PageLayout";
+import { ErrorPanel } from "components/MsgPanels";
+import { CustomTable } from "components/Table";
+
+export default class TradeBook extends React.Component {
```
Comment (szerintedmi) [inline src/containers/exchange/components/TradeBook.js]: minor nitpicking: I would call this component TradeHistory

File: src/modules/reducers/exchange.js [hunk 5]
```diff
@@ -11,7 +11,6 @@ export const EXCHANGE_REFRESH_ERROR = "exchange/EXCHANGE_REFRESH_ERROR";
 export const EXCHANGE_REFRESH_SUCCESS = "exchange/EXCHANGE_REFRESH_SUCCESS";
 
 const initialState = {
-    contract: null,
```
Comment (szerintedmi) [inline src/modules/reducers/exchange.js]: why was this removed?

File: src/containers/exchange/components/TradeBook.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+          'type': 'Type',
+          'direction': 'Direction',
+          'price': 'Price',
+          'tokenValue': 'Token Value',
+          'ethAmountRounded' : 'Eth Amount',
+          'tokenAmount': 'Token Amount'
+        }
+
+        return (
+            <Pblock loading={isLoading} header={header}>
```
Comment (szerintedmi) [inline src/containers/exchange/components/TradeBook.js]: this loading doesn't work, I can only see a Connecting... message and table headers shown.
 Also a msg insteasd of emtpy table with table headers:
https://trello.com/c/PztAuJe7/107-my-trade-history-you-have-no-trade-history-instead-of-empty-tables

File: src/modules/ethereum/tradeTransactions.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        ]);
+        const logs = await Promise.all([
+            ...logsNewOrder.map(eventLog => _formatTradeLog(NewOrder, account, eventLog)),
+            ...logsCanceledOrder.map(eventLog => _formatTradeLog(CancelledOrder, account, eventLog)),
+            ...logsOrderFillBuy.map(eventLog => _formatTradeLog(OrderFill, account, eventLog, 'buy')),
+            ...logsOrderFillSell.map(eventLog => _formatTradeLog(OrderFill, account, eventLog, 'sell'))
+        ]);
+
+        logs.sort((log1, log2) => {
+          return log2.blockData.timestamp - log1.blockData.timestamp;
```
Comment (szerintedmi) [inline src/modules/ethereum/tradeTransactions.js]: minor improvements:

1. use blockNumber for sorting (as for transfers so we don't rely on block timestamp being fetched)
1. maybe do the sort in reducer (as for transfers so it's implemented the same way and custom sorting could be passed later to the reducer ?)

File: src/modules/reducers/exchange.js (hunk 5 above)
Comment (gyurka00) [reply src/modules/reducers/exchange.js]: It's just an accidental.
I set it back.

File: src/modules/exchangeProvider.js [hunk 8]
```diff
@@ -92,7 +96,11 @@ const onCancelledOrder = function(orderId, maker, tokenAmount, weiAmount) {
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
-    store.dispatch(processNewTrade('CancelledOrder', userAccount, this));
+    if (userAccount === maker) {
+        store.dispatch(processNewTrade('CancelledOrder', userAccount, this));
+    } else {
+        processNewTrade('CancelledOrder', userAccount, this)
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: I'm not sure what are you trying to achieve here? Why are you calling processNewTrade directly without dispatching? I think this should be enough:
```js
if (userAccount === maker) {
        store.dispatch(processNewTrade('CancelledOrder', userAccount, this));
}
```
NB: I'm not sure about account capitalisation, you might need .toLowerCase() on userAccount or maker if one of them is in all lowercase ([read more about account checksum](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md))

File: src/modules/exchangeProvider.js [hunk 9]
```diff
@@ -109,7 +117,11 @@ const onOrderFill = function(tokenBuyer, tokenSeller, buyTokenOrderId, sellToken
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
-    store.dispatch(processNewTrade('OrderFill', userAccount, this));
+    if (userAccount === tokenBuyer || userAccount === tokenSeller) {
+        store.dispatch(processNewTrade('OrderFill', userAccount, this));
+    } else {
+        processNewTrade('OrderFill', userAccount, this)
+    }
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: Does it work? Have you tried it?
We want to add two trades if maker and taker are the same.
So something like this should work?
```js
if (userAccount === tokenBuyer ) {
   store.dispatch(processNewTrade('OrderFill', userAccount, this));
}

if( userAccount === tokenSeller) {
   store.dispatch(processNewTrade('OrderFill', userAccount, this));
}
```
(not sure about how you differentiate buy/sell side when parsing, you might need to pass some extra params?)

File: src/modules/exchangeProvider.js [hunk 10]
```diff
@@ -109,7 +117,11 @@ const onOrderFill = function(tokenBuyer, tokenSeller, buyTokenOrderId, sellToken
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
-    store.dispatch(processNewTrade('OrderFill', userAccount, this));
+    if (userAccount === tokenBuyer || userAccount === tokenSeller) {
+        store.dispatch(processNewTrade('OrderFill', userAccount, this));
+    } else {
+        processNewTrade('OrderFill', userAccount, this)
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: why are you calling it without dispatch?

File: src/modules/exchangeProvider.js [hunk 11]
```diff
@@ -76,7 +76,11 @@ const onNewOrder = function(orderId, maker, price, tokenAmount, weiAmount) {
     store.dispatch(refreshExchange());
     store.dispatch(refreshOrders());
     const userAccount = store.getState().web3Connect.userAccount;
-    store.dispatch(processNewTrade('NewOrder', userAccount, this));
+    if (userAccount === maker) {
+        store.dispatch(processNewTrade('NewOrder', userAccount, this));
+    } else {
+        processNewTrade('NewOrder', userAccount, this)
+    }
```
Comment (szerintedmi) [inline src/modules/exchangeProvider.js]: same applies to this as for CancelledOrder below

File: src/modules/exchangeProvider.js (hunk 8 above)
Comment (gyurka00) [reply src/modules/exchangeProvider.js]: okay, I understand and I made a fix.
